src/main/host/descriptor/socket/unix.rs

```diff
@@ -0,0 +1,400 @@
+                    let mut socket = socket.borrow_mut();
+
+                    // if the socket is already closed, do nothing
+                    if socket.state.contains(FileState::CLOSED) {
+                        return;
+                    }
+
+                    // the socket is readable iff the buffer is readable
+                    socket.copy_state(FileState::READABLE, state, event_queue);
+                }
+            },
+        );
+
+        socket_ref.recv_buffer_event_handle = Some(recv_handle);
+
+        std::mem::drop(socket_ref);
+
+        socket
+    }
+
+    pub fn get_status(&self) -> FileStatus {
+        self.status
+    }
+
+    pub fn set_status(&mut self, status: FileStatus) {
+        self.status = status;
+    }
+
+    pub fn mode(&self) -> FileMode {
+        self.mode
+    }
+
+    pub fn get_bound_address(&self) -> Option<nix::sys::socket::UnixAddr> {
+        self.bound_addr
+    }
+
+    pub fn set_bound_address(
+        &mut self,
+        addr: Option<nix::sys::socket::UnixAddr>,
+    ) -> Result<(), SyscallError> {
+        // if already bound
+        if self.bound_addr.is_some() {
+            // TODO: not sure what should happen here
+            return Err(Errno::EINVAL.into());
+        }
+
+        self.bound_addr = addr;
+        Ok(())
+    }
+
+    pub fn get_peer_address(&self) -> Option<nix::sys::socket::UnixAddr> {
+        self.peer_addr
+    }
+
+    pub fn address_family(&self) -> nix::sys::socket::AddressFamily {
+        nix::sys::socket::AddressFamily::Unix
+    }
+
+    pub fn recv_buffer(&self) -> &Arc<AtomicRefCell<SharedBuf>> {
+        &self.recv_buffer
+    }
+
+    pub fn close(&mut self, event_queue: &mut EventQueue) -> SyscallResult {
+        // drop the event listener handles so that we stop receiving new events
+        self.send_buffer_event_handle
+            .take()
+            .map(|h| h.stop_listening());
+        self.recv_buffer_event_handle
+            .take()
+            .map(|h| h.stop_listening());
+
+        // inform the buffer that there is one fewer writers
+        if let Some(send_buffer) = self.send_buffer.as_ref() {
+            send_buffer.borrow_mut().remove_writer(event_queue);
+        }
+
+        // no need to hold on to the send buffer anymore
+        self.send_buffer = None;
+
+        // set the closed flag and remove the active, readable, and writable flags
+        self.copy_state(
+            FileState::CLOSED | FileState::ACTIVE | FileState::READABLE | FileState::WRITABLE,
+            FileState::CLOSED,
+            event_queue,
+        );
+
+        Ok(0.into())
+    }
+
+    pub fn read<W>(
+        &mut self,
+        mut _bytes: W,
+        _offset: libc::off_t,
+        _event_queue: &mut EventQueue,
+    ) -> SyscallResult
+    where
+        W: std::io::Write + std::io::Seek,
+    {
+        // we could call recvfrom() here, but for now we expect that there are no code paths that
```

Comment:
Would application-code calling `read` on the socket get here? If so, seems like we ought to handle it more gracefully (though OK if not in this PR)

Reply:
The `read` syscall handler automatically calls the `recvfrom` syscall handler if the fd is a socket, so calling `read()` on a socket fd should never reach here.

Reply:
Expanded the comment.